nemo/collections/tts/modules/common.py

```diff
@@ -0,0 +1,869 @@
+        self.dropout = nn.Dropout(p=p_dropout)
+
+        convolutions = []
+        for i in range(n_layers):
+            conv_layer = ConvNorm(
+                in_dim if i == 0 else n_channels,
+                n_channels,
+                kernel_size=kernel_size,
+                stride=1,
+                padding=int((kernel_size - 1) / 2),
+                dilation=1,
+                w_init_gain='relu',
+            )
+            conv_layer = torch.nn.utils.weight_norm(conv_layer.conv, name='weight')
+            convolutions.append(conv_layer)
+
+        self.convolutions = nn.ModuleList(convolutions)
+
+        self.bilstm = nn.LSTM(n_channels, int(n_channels // 2), 1, batch_first=True, bidirectional=True)
+        lstm_norm_fn_pntr = nn.utils.spectral_norm
+        self.bilstm = lstm_norm_fn_pntr(self.bilstm, 'weight_hh_l0')
+        self.bilstm = lstm_norm_fn_pntr(self.bilstm, 'weight_hh_l0_reverse')
+        self.dense = nn.Linear(n_channels, out_dim)
+
+    def run_padded_sequence(self, context, lens):
+        context_embedded = []
+        for b_ind in range(context.size()[0]):  # TODO: speed up
+            curr_context = context[b_ind : b_ind + 1, :, : lens[b_ind]].clone()
+            for conv in self.convolutions:
+                curr_context = self.dropout(F.relu(conv(curr_context)))
+            context_embedded.append(curr_context[0].transpose(0, 1))
+        context = torch.nn.utils.rnn.pad_sequence(context_embedded, batch_first=True)
+        return context
+
+    def run_unsorted_inputs(self, fn, context, lens):
+        lens_sorted, ids_sorted = torch.sort(lens, descending=True)
+        unsort_ids = [0] * lens.size(0)
+        for i in range(ids_sorted.shape[0]):
+            unsort_ids[ids_sorted[i]] = i
+        lens_sorted = lens_sorted.long().cpu()
+
+        context = context[ids_sorted]
+        context = nn.utils.rnn.pack_padded_sequence(context, lens_sorted, batch_first=True)
+        context = fn(context)[0]
+        context = nn.utils.rnn.pad_packed_sequence(context, batch_first=True)[0]
+
+        # map back to original indices
+        context = context[unsort_ids]
+        return context
+
+    def forward(self, context, lens):
+        if context.shape[0] > 1:
+            context = self.run_padded_sequence(context, lens)
+        else:
+            for conv in self.convolutions:
+                context = self.dropout(F.relu(conv(context)))
+            context = context.transpose(1, 2)
+
+        self.bilstm.flatten_parameters()
+        if lens is not None:
+            context = self.run_unsorted_inputs(self.bilstm, context, lens)
+        else:
+            context = self.bilstm(context)[0]
+
+        x_hat = self.dense(context).permute(0, 2, 1)
+
+        return x_hat
+
+    def infer(self, z, txt_enc, spk_emb):
+        x_hat = self.forward(txt_enc, spk_emb)['x_hat']
+        x_hat = self.feature_processing.denormalize(x_hat)
+        return x_hat
+
+
+class Encoder(nn.Module):
```

Comment:
Are all of those RadTTS submodules belong to TTS common.py ? Even if so - Encoder should be renamed to something more specific.

Reply:
changed to RADTTSENCODER. Thanks